plutus-tx/src/PlutusTx/Data.hs

```diff
@@ -90,34 +88,91 @@ the indefinite kinds.
-    -- See Note [Definite and indefinite forms of CBOR]
-    CBOR.TMap es                              -> Map <$> traverse (\(t1, t2) -> (,) <$> fromTerm t1 <*> fromTerm t2) es
-    CBOR.TMapI es                             -> Map <$> traverse (\(t1, t2) -> (,) <$> fromTerm t1 <*> fromTerm t2) es
-    CBOR.TList l                              -> List <$> traverse fromTerm l
-    CBOR.TListI l                             -> List <$> traverse fromTerm l
-    CBOR.TInteger i                           -> pure $ I i
-    CBOR.TInt i                               -> pure $ I $ fromIntegral i
-    CBOR.TBytes b                             -> if BS.length b <= 64
-                                                   then pure $ B b
-                                                   else throwError "ByteString exceeds 64"
-    CBOR.TBytesI b                            -> if BSL.length b <= 64
-                                                   then pure $ B $ BSL.toStrict b
-                                                   else throwError "ByteString exceeds 64"
-    _                                         -> throwError "Unsupported kind of CBOR"
-
--- See Note [Definite and indefinite forms of CBOR]
--- | View a CBOR Term as an Integer if possible.
-asInt :: CBOR.Term -> Maybe Integer
-asInt (CBOR.TInteger i) = Just i
-asInt (CBOR.TInt i)     = Just $ fromIntegral i
-asInt _                 = Nothing
+decodeData :: forall s. Decoder s Data
+decodeData = CBOR.peekTokenType >>= \case
+  CBOR.TypeUInt         -> decodeInt
+  CBOR.TypeUInt64       -> decodeInt
+  CBOR.TypeNInt         -> decodeInt
+  CBOR.TypeNInt64       -> decodeInt
+  CBOR.TypeInteger      -> decodeBigInt
+
+  CBOR.TypeBytes        -> decodeBytes
+  CBOR.TypeBytesIndef   -> decodeBytes
+
+  CBOR.TypeListLen      -> decodeList
+  CBOR.TypeListLen64    -> decodeList
+  CBOR.TypeListLenIndef -> decodeList
+
+  CBOR.TypeMapLen       -> decodeMap
+  CBOR.TypeMapLen64     -> decodeMap
+  CBOR.TypeMapLenIndef  -> decodeMap
+
+  CBOR.TypeTag          -> decodeConstr
+  CBOR.TypeTag64        -> decodeConstr
+
+  _                     -> fail "Unrecognized value"
+
+decodeInt :: Decoder s Data
+decodeInt = I <$> CBOR.decodeInteger
+
+decodeBigInt :: Decoder s Data
+decodeBigInt = do
+  convert <- CBOR.decodeTag64 >>= \case
+    2 -> pure Magic.uintegerFromBytes
+    3 -> pure Magic.nintegerFromBytes
+    _ -> fail "Impossible! Unreachable case was reached."
+  bytes <- decodeBytesBounded
+  pure $ I (convert bytes)
+
+decodeBytes :: Decoder s Data
+decodeBytes = B <$> decodeBytesBounded
+
+decodeBytesBounded :: Decoder s BS.ByteString
+decodeBytesBounded =  do
+  b <- CBOR.decodeBytes
+  if BS.length b <= 64
+    then pure b
+    else fail "ByteString exceeds 64 bytes"
+
+decodeList :: Decoder s Data
+decodeList = List <$> decodeListOf decodeData
+
+decodeListOf :: Decoder s x -> Decoder s [x]
```

Comment:
There aren't list decoder functions from somewhere we can borrow? Things like this seem fairly generic...

Reply:
At the moment the ledger has a bunch of utilities like this.

They should probably live in cardano-base

Reply:
Or `cborg`!

Reply:
That would be even better :)